GraphQL documentation (#2392)
elastic/connectors: docs/reference/graphql.md

# GraphQL Connector

The [Elastic GraphQL connector](../connectors/sources/graphql.py) is built with the Elastic connectors Python framework and is available as a self-managed [connector client](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html).

## Availability and prerequisites

This connector is available as a **connector client** from the **Python connectors framework**. To use this connector, satisfy all [connector client requirements](https://www.elastic.co/guide/en/enterprise-search/master/build-connector.html).

> This connector is in **beta** and is subject to change. The design and code is less mature than official GA features and is being provided as-is with no warranties. Beta features are not subject to the support SLA of official GA features.

Reviewer: tech preview

## Usage

To use this connector as a **connector client**, select the **GraphQL** tile when creating a new connector under **Search -> Connectors**.

For additional operations, see [Usage](https://www.elastic.co/guide/en/enterprise-search/master/connectors-usage.html).

## Configuration

### Configure GraphQL connector

Note the following configuration fields:

#### `http_endpoint`  (required)

Provide base URL of the source with endpoint. Example:

 `https://api.xyz.com/graphql`

#### `http_method`  (required)

> Select method from GET and POST.

Reviewer: ```suggestion
Select method from `GET` and `POST`.
```

In case of `Basic Auth` authentication method, provides password to connect with the source.

#### `token` (required in case of `Bearer Token`)

In case of `Bearer token` authentication method, provides token to connect with the source.

#### `graphql_query` (required)

This query will be used to fetch data from the source. The query can contain variables which can be provided in the `graphql_variables` field. The connector will substitute the variables in the query with values from `graphql_variables` and make a GraphQL query to the source. Example:

```
query getUser($id: ID!) { 
    user(id: $id) {
        name 
        email 
    }
}
```

#### `graphql_variables`

A json of key value pairs of variables used in the GraphQL query. The connector will substitute the variables in the query with the values provided here and make a GraphQL query to the source.

Example: If the GraphQL query is `query getUser($id: ID!) { user(id: $id) { name } }` and the value of `graphql_variables` is `{"id": "123"}`, the final query will be `query getUser { user(id: "123") { name } }` and the connector will make a GraphQL query to the source with the modified query.

> #### `graphql_object_list` (required)

Reviewer: Ahh this variable name slipped through the review, shold we rename it to sth like `graphql_object_id_mapping`

Author: As we need to change this throughout the entire connector (graphql.py, test cases, and E2E), would it be okay if I make this change in a follow-up PR?

Reviewer: Added the followup ticket https://github.com/elastic/connectors/issues/2420

> Provides json of GraphQL object types to fetch data from. The connector will fetch data for each object type in this json. The GraphQL objects can be nested and the connector will recursively fetch data for nested objects. The connector will fetch all the fields for each object type. The value of each key in the json should be a name of unique field. Use '.' to provide full path of the object from the root of the response.  Example:

Reviewer: ```suggestion
A JSON mapping between GraphQL response objects to index and their ID fields. The connector will fetch data for each object (JSON key) and use the provided ID field (JSON value) to index the object into Elasticsearch. The connector will index all fields for each object specified in the mapping. Use a dot `(.)` notation to specify the full path from the root of the GraphQL response to the desired object.

Example: 
```

```
If the GraphQL query is `query getUser { organization { users{ user_id name email} } }` which is fetching all available users from the source and we want to index every user as separate document in that case configure this field as below.
{
    "organization.users": "user_id"
}

Here user_id is unique in all the user document so configure user_id as a value to `organization.users`.
```

> Note: The path provided in this field should only contain Json Objects and not lists.

Reviewer: This not needed if we improve the `graphql_object_list` name

#### `headers`

Json of custom headers to be sent with each GraphQL request. Example:

```
{
    "content-type": "Application/json"
}
```

#### `pagination_model` (required)

> This field specifies the pagination model to be used by the connector. The connector supports `No pagination` and `Cursor-based pagination` pagination models. For cursor-based pagination, add `pageInfo` and an `after` argument variable in your query at the desired node (Pagination key). Use `after` query argument with a variable to iterate through pages. Detailed examples and setup instructions are available in the docs. The default value for this field is `No pagination`.

Reviewer: We def need an example query to explain how this works

# GraphQL Connector

> The [Elastic GraphQL connector](../connectors/sources/graphql.py) is built with the Elastic connectors Python framework and is available as a self-managed [connector client](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html).

Reviewer 2: ```suggestion
The Elastic GraphQL connector connector is written in Python using the [Elastic connector framework](https://github.com/elastic/connectors/tree/main). View the [source code for this connector](https://github.com/elastic/connectors/blob/main/connectors/sources/graphql.py).
```

## Availability and prerequisites

> This connector is available as a **connector client** from the **Python connectors framework**. To use this connector, satisfy all [connector client requirements](https://www.elastic.co/guide/en/enterprise-search/master/build-connector.html).

Reviewer 2: ```suggestion
This connector is available as a self-managed connector client.
To use this connector, satisfy all [connector client requirements](https://www.elastic.co/guide/en/enterprise-search/master/build-connector.html).
```

For `Cursor-based pagination`, the query should look like below example:
```
query getUsers($cursor: String!) {
    sampleData { 
        users(after: $cursor) {
            pageInfo {
                endCursor
                hasNextPage
            }
            nodes {
                first_name
                last_name
                address
            } 
        } 
    } 
}
```
and the value of `pagination_key` is `sampleData.users` so, it must contain `pageInfo {endCursor hasNextPage}` and `after` argument with a variable in case of `Cursor-based pagination`.

#### `pagination_key` (required)

Specifies which GraphQL object is used for pagination. Use `.` to provide full path of the object from the root of the response. Example:

- `organization.users`

#### `connection_timeout`

The `connection_timeout` specifies the maximum time in seconds to wait for a response from the GraphQL source. Default value is 30 seconds.

## Documents and syncs

> The connector syncs the objects and entities based on GraphQL Query and GraphQL Object List.

Reviewer 2: ```suggestion
The connector syncs the objects and entities based on GraphQL Query and GraphQL Object List.

### Sync types
[Full syncs](https://www.elastic.co/guide/en/enterprise-search/current/connectors-sync-types.html#connectors-sync-types-full) are supported by default for all connectors.

This connector currently does not support [incremental syncs](https://www.elastic.co/guide/en/enterprise-search/current/connectors-sync-types.html#connectors-sync-types-incremental).
```

```
query getUser($id: ID!) { 
    user(id: $id) {
        name 
        email 
    }
}
```

#### `graphql_variables`

A json of key value pairs of variables used in the GraphQL query. The connector will substitute the variables in the query with the values provided here and make a GraphQL query to the source.

Example: If the GraphQL query is `query getUser($id: ID!) { user(id: $id) { name } }` and the value of `graphql_variables` is `{"id": "123"}`, the final query will be `query getUser { user(id: "123") { name } }` and the connector will make a GraphQL query to the source with the modified query.

#### `graphql_object_list` (required)

A JSON mapping between GraphQL response objects to index and their ID fields. The connector will fetch data for each object (JSON key) and use the provided ID field (JSON value) to index the object into Elasticsearch. The connector will index all fields for each object specified in the mapping. Use a dot `(.)` notation to specify the full path from the root of the GraphQL response to the desired object.

Example: 

```

> If the GraphQL query is `query getUser { organization { users{ user_id name email} } }` which is fetching all available users from the source and we want to index every user as separate document in that case configure this field as below.

Reviewer: This comment should be outside of the code block, it will render as a one-line that is hard to read

{
    "organization.users": "user_id"
}

> Here user_id is unique in all the user document so configure user_id as a value to `organization.users`.

Reviewer: Probz should be outside of the code block

```

Note: The path provided in this field should only contain Json Objects and not lists.

#### `headers`

Json of custom headers to be sent with each GraphQL request. Example:

```
{
    "content-type": "Application/json"
}
```

#### `pagination_model` (required)

This field specifies the pagination model to be used by the connector. The connector supports `No pagination` and `Cursor-based pagination` pagination models. For cursor-based pagination, add `pageInfo {endCursor hasNextPage}` and an `after` argument variable in your query at the desired node (`Pagination key`). Use `after` query argument with a variable to iterate through pages. Detailed examples and setup instructions are available in the docs. The default value for this field is `No pagination`. Example:

```

> For `Cursor-based pagination`, the query should look like below example:

Reviewer: This should be outside of the code block

query getUsers($cursor: String!) {
    sampleData { 
        users(after: $cursor) {
            pageInfo {
                endCursor
                hasNextPage
            }
            nodes {
                first_name
                last_name
                address
            } 
        } 
    } 

> }

Reviewer: your forgot to add ``` somewhere here

> and the value of `pagination_key` is `sampleData.users` so, it must contain `pageInfo {endCursor hasNextPage}` and `after` argument with a variable in case of `Cursor-based pagination`.

Reviewer: this should be outside of the code block

Reviewer: the whole section is inside a code block

Author: Oh yes, thank you for pointing that out.

#### `pagination_key` (required)

Specifies which GraphQL object is used for pagination. Use '.' to provide full path of the object from the root of the response. Example:

> - 'organization.users'

Reviewer: inconsistent use of `'` vs ` in couple of places here
